core/src/main/scala/kafka/controller/ReplicaStateMachine.scala

```diff
@@ -122,178 +139,214 @@ class ReplicaStateMachine(controller: KafkaController, stateChangeLogger: StateC
+    val getDataResponses = try {
+      zkUtils.getTopicPartitionStatesRaw(partitions)
+    } catch {
+      case e: Exception =>
+        partitions.foreach(partition => failed.put(partition, e))
+        return (leaderAndIsrs.toMap, partitionsWithNoLeaderAndIsrInZk, failed.toMap)
+    }
+    getDataResponses.foreach { getDataResponse =>
+      val partition = getDataResponse.ctx.asInstanceOf[TopicAndPartition]
+      if (Code.get(getDataResponse.rc) == Code.OK) {
+        val leaderIsrAndControllerEpochOpt = TopicPartitionStateZNode.decode(getDataResponse.data, getDataResponse.stat)
+        if (leaderIsrAndControllerEpochOpt.isEmpty) {
+          partitionsWithNoLeaderAndIsrInZk += partition
+        } else {
+          val leaderIsrAndControllerEpoch = leaderIsrAndControllerEpochOpt.get
+          if (leaderIsrAndControllerEpoch.controllerEpoch > controllerContext.epoch) {
+            val exception = new StateChangeFailedException("Leader and isr path written by another controller. This probably" +
+              s"means the current controller with epoch ${controllerContext.epoch} went through a soft failure and another " +
+              s"controller was elected with epoch ${leaderIsrAndControllerEpoch.controllerEpoch}. Aborting state change by this controller")
+            failed.put(partition, exception)
+          } else {
+            leaderAndIsrs.put(partition, leaderIsrAndControllerEpoch.leaderAndIsr)
+          }
+        }
+      } else if (Code.get(getDataResponse.rc) == Code.NONODE) {
+        partitionsWithNoLeaderAndIsrInZk += partition
+      } else {
+        failed.put(partition, KeeperException.create(Code.get(getDataResponse.rc)))
+      }
+    }
+    (leaderAndIsrs.toMap, partitionsWithNoLeaderAndIsrInZk, failed.toMap)
   }
 
   def isAtLeastOneReplicaInDeletionStartedState(topic: String): Boolean = {
-    val replicasForTopic = controller.controllerContext.replicasForTopic(topic)
+    val replicasForTopic = controllerContext.replicasForTopic(topic)
     val replicaStatesForTopic = replicasForTopic.map(r => (r, replicaState(r))).toMap
     replicaStatesForTopic.foldLeft(false)((deletionState, r) => deletionState || r._2 == ReplicaDeletionStarted)
   }
 
   def replicasInState(topic: String, state: ReplicaState): Set[PartitionAndReplica] = {
-    replicaState.filter(r => r._1.topic.equals(topic) && r._2 == state).keySet
+    replicaState.filter(r => r._1.topic.equals(topic) && r._2 == state).keySet.toSet
   }
 
-  def isAnyReplicaInState(topic: String, state: ReplicaState): Boolean = {
-    replicaState.exists(r => r._1.topic.equals(topic) && r._2 == state)
+  def areAllReplicasForTopicDeleted(topic: String): Boolean = {
+    val replicasForTopic = controllerContext.replicasForTopic(topic)
+    val replicaStatesForTopic = replicasForTopic.map(r => (r, replicaState(r))).toMap
+    debug(s"Are all replicas for topic $topic deleted $replicaStatesForTopic")
+    replicaStatesForTopic.forall(_._2 == ReplicaDeletionSuccessful)
```

Comment:
Could we use case to avoid unnamed reference _._2? There are a few other places like that in this file.